### PR TITLE
Support tencent COS fileIO

### PR DESCRIPTION
**Pull request by lubo212**
Tencent COS is a popular object storage service in public cloud services, and this PR provides support for it.

### CODE REVIEW

File: tencent/src/test/java/org/apache/iceberg/tencentcloud/cos/TestCosClientFactories.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+ * under the License.
+ */
+package org.apache.iceberg.tencentcloud.cos;
+
+import com.qcloud.cos.COS;
+import java.util.Map;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.junit.Assert;
+import org.junit.Test;
```
Comment (nastra) [inline tencent/src/test/java/org/apache/iceberg/tencentcloud/cos/TestCosClientFactories.java:26]: Can you please update all tests to JUnit5 + AssertJ-style checks. The codebase is currently migrating away from JUnit4. See also https://iceberg.apache.org/contribute/#testing

Comment (lubo212) [reply tencent/src/test/java/org/apache/iceberg/tencentcloud/cos/TestCosClientFactories.java]: I think I have finished adjusting. Please help me check.

File: tencent/src/test/java/org/apache/iceberg/tencentcloud/cos/TestCosFileIO.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.io.ByteStreams;
+import org.apache.iceberg.util.SerializableSupplier;
+import org.apache.iceberg.util.SerializationUtil;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.*;
```
Comment (nastra) [inline tencent/src/test/java/org/apache/iceberg/tencentcloud/cos/TestCosFileIO.java:42]: * imports aren't allowed, so it's better to import `assertThat` directly here

Comment (nastra) [reply tencent/src/test/java/org/apache/iceberg/tencentcloud/cos/TestCosFileIO.java:42]: I've also noticed that formatting is off in a few places, you might want to run `./gradlew spotlessApply` locally to fix formatting

**Review by nastra — COMMENTED** [review]
I did a high-level pass over the code and left a few initial comments

File: build.gradle [hunk 3]
```diff
@@ -948,6 +948,48 @@ project(':iceberg-snowflake') {
...
+  dependencies {
+    implementation project(path: ':iceberg-bundled-guava', configuration: 'shadow')
+    api project(':iceberg-api')
+    implementation project(':iceberg-core')
+    implementation project(':iceberg-common')
+
+    compileOnly (libs.cos.sdk) {
+      exclude group: 'com.google.code.gson', module: 'gson'
+    }
+    compileOnly (libs.cos.sdk.bundle) {
```
Comment (nastra) [inline build.gradle:965]: is the SDK bundle really required?

File: build.gradle [hunk 4]
```diff
@@ -948,6 +948,48 @@ project(':iceberg-snowflake') {
...
+    compileOnly(libs.hadoop2.common) {
+      exclude group: 'org.apache.avro', module: 'avro'
+      exclude group: 'org.slf4j', module: 'slf4j-log4j12'
+      exclude group: 'javax.servlet', module: 'servlet-api'
+      exclude group: 'com.google.code.gson', module: 'gson'
+    }
+
+    testImplementation libs.jackson.dataformat.xml
+    testImplementation project(path: ':iceberg-api', configuration: 'testArtifacts')
+    testImplementation libs.spring.web
```
Comment (nastra) [inline build.gradle:980]: why are all these spring dependencies defined here? I haven't seen any usage of those in the tests. I'd suggest to remove dependencies that aren't necessary and slim this down to the absolute minimum

File: build.gradle [hunk 5]
```diff
@@ -948,6 +948,48 @@ project(':iceberg-snowflake') {
...
+    implementation project(':iceberg-core')
+    implementation project(':iceberg-common')
+
+    compileOnly (libs.cos.sdk) {
+      exclude group: 'com.google.code.gson', module: 'gson'
+    }
+    compileOnly (libs.cos.sdk.bundle) {
+      exclude group: 'com.google.code.gson', module: 'gson'
+    }
+    compileOnly libs.jaxb.api
```
Comment (nastra) [inline build.gradle:968]: are these required? I see that they have been copy/pasted from the `':iceberg-aliyun'`  module would I'd suggest to slim this down to what is required for this project

**Pull request closed without merging**